alns/select/SelectionScheme.py

```diff
@@ -0,0 +1,101 @@
+from abc import ABC, abstractmethod
+from typing import Tuple, Optional
+
+import numpy as np
+from numpy.random import RandomState
+
+
+class SelectionScheme(ABC):
+    """
+    Base class from which to implement an operator selection scheme.
+
+    Parameters
+    ----------
+    num_destroy
+        Number of destroy operators.
+    num_repair
+        Number of repair operators.
+    op_coupling
+        Optional keyword argument. Matrix that indicates coupling between
+        destroy and repair operators. Entry (i, j) is 1 if destroy operator i
+        can be used in conjunction with repair operator j and 0 otherwise.
+    """
+
+    def __init__(
+        self,
+        num_destroy: int,
+        num_repair: int,
+        *,
+        op_coupling: Optional[np.ndarray] = None,
+    ):
+        self._validate_arguments(num_destroy, num_repair, op_coupling)
+
+        self._num_destroy = num_destroy
+        self._num_repair = num_repair
+        self._op_coupling = (
+            op_coupling
+            if op_coupling is not None
+            else np.ones((num_destroy, num_repair))
+        )
+
+    @property
+    def num_destroy(self) -> int:
+        return self._num_destroy
+
+    @property
+    def num_repair(self) -> int:
+        return self._num_repair
+
+    @property
+    def op_coupling(self) -> np.ndarray:
+        return self._op_coupling
+
+    @abstractmethod
+    def select_operators(self, rnd_state: RandomState) -> Tuple[int, int]:
+        """
+        Selects a destroy and repair operator pair to apply in this iteration.
+
+        Parameters
+        ----------
+        rnd_state
+            Random state object, to be used for random number generation.
+
+        Returns
+        -------
+        A tuple of (d_idx, r_idx), which are indices into the destroy and
+        repair operator lists, respectively.
+        """
+        return NotImplemented
+
+    @abstractmethod
+    def update(self, d_idx: int, r_idx: int, s_idx: int):
```

Comment:
If we have access to curr/cand/best in `SelectionScheme.update`, then we can also get rid of all the outcomes `_BEST`, `_BETTER`, etc. that are now being determined in ALNS. Those are selection scheme-specific anyway. The only thing that's not trivial is to determine the `_ACCEPT` outcome. What do you think?

Reply:
Agree; also see other comment. This is a breaking change, so this will require a bump to v5.

We might want to toy implementing some more advanced selection schemes too, while we're at it. But not in this PR.